lib/pubsubstub/subscriber.rb

```diff
@@ -0,0 +1,88 @@
+module Pubsubstub
+  class Subscriber
+    include Logging
+    include Mutex_m
+
+    def initialize
+      super
+      @subscribed = false
+      @listeners = {}
+    end
+
+    def subscribed?
+      @subscribed
+    end
+
+    def add_event_listener(channel_key, callback)
+      synchronize do
+        @listeners[channel_key] ||= Set.new
+        !!@listeners[channel_key].add?(callback)
+      end
+    end
+
+    def remove_event_listener(channel_key, callback)
+      synchronize do
+        return unless @listeners[channel_key]
+        !!@listeners[channel_key].delete?(callback)
+      end
+    end
+
+    def stop
+      # redis.client.call allow to bypass the client mutex
+      # Since we now that the only other possible caller is blocking on reading the socket this is safe
+      synchronize do
+        redis.client.call(['punsubscribe', pubsub_pattern])
+      end
+    end
+
+    def start
+      redis.psubscribe(pubsub_pattern) do |on|
+        on.psubscribe do
+          info { "Subscribed to #{pubsub_pattern}" }
+          @subscribed = true
+        end
+
+        on.punsubscribe do
+          info { "Unsubscribed from #{pubsub_pattern}" }
+          @subscribed = false
+        end
+
+        on.pmessage do |pattern, pubsub_key, message|
+          process_message(pubsub_key, message)
+        end
+      end
+    ensure
+      info { "Terminated" }
+    end
+
+    def pubsub_pattern
+      '*.pubsub'
+    end
+
+    def process_message(pubsub_key, message)
```

Comment:
tbh I'd rather have a `send` in tests than have a public interface just for tests. would you mind changing it? I'll merge after